base/readiness.go

```diff
@@ -0,0 +1,197 @@
+package base
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	log "github.com/iter8-tools/iter8/base/log"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/util/retry"
+)
+
+const (
+	// ReadinessTaskName is the task name
+	ReadinessTaskName = "k8s-object-ready"
+
+	// defaultTimeout is default timeout for readiness command
+	defaultTimeout = "10s"
+)
+
+// ReadinessInputs identifies the K8s object to test for existence and
+// the (optional) condition that should be tested (succeeds if true).
+type readinessInputs struct {
+	// APIVersion of the object. Optional. If unspecified it will be defaulted to ""
+	APIVersion string `json:"apiVersion,omitempty" yaml:"apiVersion,omitempty"`
+	// Resource type of the object.
+	Resource string `json:"resource" yaml:"resource"`
+	// Namespace of the object. Optional. If left unspecified, this will be defaulted to the namespace of the experiment
+	Namespace *string `json:"namespace,omitempty" yaml:"namespace,omitempty"`
+	// Name of the object
+	Name string `json:"name" yaml:"name"`
+	// Condition is label of condition to check for value of "True"
+	Condition *string `json:"condition" yaml:"condition"`
+	// Timeout is maximum time spent trying to find object and check condition
+	Timeout *string `json:"timeout" yaml:"timeout"`
+}
+
+// ReadinessTask checks existence and readiness of specified resources
+type readinessTask struct {
+	TaskMeta
+	With readinessInputs `json:"with" yaml:"with"`
+}
+
+// initializeDefaults sets default values for the readiness task
+func (t *readinessTask) initializeDefaults() {
+	if t.With.Timeout == nil {
+		t.With.Timeout = StringPointer(defaultTimeout)
+	}
+}
+
+// validateInputs validates task inputs
+// at present all validation at initialization
+func (t *readinessTask) validateInputs() error {
+	return nil
+}
+
+// run executes the task
+func (t *readinessTask) run(exp *Experiment) error {
+	// validation
+	err := t.validateInputs()
+	if err != nil {
+		return err
+	}
+
+	// initialization
+	t.initializeDefaults()
+	kd.initKube()
+	// set Namespace (from context) if not already set
+	if t.With.Namespace == nil {
```

Comment:
Move this into initializeDefaults()... 
```go
if t.With.Namespace == nil {
  t.With.Namespace = StringPointer(kd.Namespace())
}
```

Reply:
this depends on cluster initialization; moved that as well.